### PR TITLE
aml: add version 0.2.0

### PR DESCRIPTION
**Pull request by NicolasDenoyelle**
This package manages variants and dependencies for installing
aml with support for multiple backends and dependencies.

### CODE REVIEW

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
The build of your PR fails because you added
```py
    url = "https://github.com/anlsys/aml/archive/refs/tags/v0.2.0.tar.gz"
```
and since this is not the release tarball (which has the generated `configure` script)
but an archive of the `git` tag, this does not have the ```configure``` script.

This means the `spack` install would have to run `autoreconf` to generate it but you didn't add the needed dependencies for it.

Instead, use
```py
    url         = 'https://github.com/anlsys/aml/releases/download/v0.2.0/aml-0.2.0.tar.gz'
```
Note: The url variable is global for the recipe/class, it's not once for each version.
spack converts the `0.2.0` in it to other versions when you build `0.1.0`, therefore
it does not belong to the specific version. I've a commit which fixes these issues below.

Normally I push such changes to your the submitted branch,
but it looks like you/your repo disables this.

Since you/your branch disables permission to push to it for
this PR, please apply this commit using: git am <filename>
This is the commit/patch:
```diff
From 70a0a9c3b5113bdc7d334ef481b34aff3595b092 Mon Sep 17 00:00:00 2001
From: Bernhard Kaindl <bernhardkaindl7@gmail.com>
Date: Thu, 2 Dec 2021 22:59:38 +0000
Subject: [PATCH 1/1] fix the build of 0.2.0 and bring some sense of order

---
 .../repos/builtin/packages/aml/package.py     | 27 +++----------------
 1 file changed, 3 insertions(+), 24 deletions(-)

diff --git a/var/spack/repos/builtin/packages/aml/package.py b/var/spack/repos/builtin/packages/aml/package.py
index 18fcf87f63..0f79ff9872 100644
--- a/var/spack/repos/builtin/packages/aml/package.py
+++ b/var/spack/repos/builtin/packages/aml/package.py
@@ -10,8 +10,6 @@ class Aml(AutotoolsPackage):
     """
     AML: Building Blocks for Memory Management.
 
-    For information on AML please visit AML homepage.
-
     Possible variants for aml spack package are:
     - `+cuda`: enables library features implementation over cuda runtime library.
     All cuda variants from the spack package "cuda" are allowed.
@@ -40,25 +38,19 @@ class Aml(AutotoolsPackage):
     - spack install aml +ze openmp-targets=spir64
     """
 
-    homepage = "https://argo-aml.readthedocs.io/"
-
+    homepage    = 'https://github.com/anlsys/aml'
+    url         = 'https://github.com/anlsys/aml/releases/download/v0.2.0/aml-0.2.0.tar.gz'
     maintainers = ['perarnau']
-
     test_requires_compiler = True
-
     tags = ['e4s']
 
     # Package sources
     ###################################
 
-    # Disable git version.
     # git = "https://github.com/anlsys/aml.git"
     # version('master', branch='master', submodules=True)
-
-    url = "https://www.mcs.anl.gov/research/projects/argo/downloads/aml-0.1.0.tar.gz"
     version('0.1.0', sha256='cc89a8768693f1f11539378b21cdca9f0ce3fc5cb564f9b3e4154a051dcea69b')
-    url = "https://github.com/anlsys/aml/archive/refs/tags/v0.2.0.tar.gz"
-    version('0.2.0', sha256='eea44780c11cff3f585f7ad1bc4fcc0d9d46aa8c27de234f6ae3703c9bb57684')
+    version('0.2.0', sha256='2044a2f3f1d7a19827dd9c0726172b690189b4d3fe938656c4160c022468cc4a')
 
     # Variants and configure options management
     ###########################################
```

**Comment by bernhardkaindl** [comment]
I converted this to a draft to indicate that this recipe needs to be fixed/cleaned up before approval.

After you applied my requested change, also cleanup the commented code for building `@master` if you really do not want to build the master branch from spack and also cleanup the test if you don't want the test.

BTW, why remove it and @master?

**Comment by NicolasDenoyelle** [comment]
@bernhardkaindl thank you for the input.
So the git archive does not appear to have a release archive with a configure either.
We are going to have to host the distribution release archieve on our website as we did for the previous version. I will run this by my colleagues and update you.

**Comment by bernhardkaindl** [comment]
@NicolasDenoyelle :
> So the git archive does not appear to have a release archive with a configure either.
The git archive files doe not have the configure, but these files have the configure scripts:
```py
wget 'https://github.com/anlsys/aml/releases/download/v0.2.0/aml-0.2.0.tar.gz'
tar tvf aml-0.2.0.tar.gz | grep 0.2.0/configure
-rwxr-xr-x 1000/100     559770 2021-12-02 03:22 aml-0.2.0/configure
-rw-r--r-- 1000/100      13715 2021-12-02 03:14 aml-0.2.0/configure.ac
```
This is different from `https://github.com/anlsys/aml/archive/refs/tags/v0.2.0.tar.gz`, which does not have a auto-generated `configure` script.
> We are going to have to host the distribution release archieve on our website as we did for the previous version.

No need, @perarnau just fixed it (mostly), but you have to update the sha256 of 0.2.0 for the sha256 of the release tarball with the configure:
```py
==> [2021-12-03-05:48:03.481690] Fetching https://github.com/anlsys/aml/releases/download/v0.2.0/aml-0.2.0.tar.gz
==> [2021-12-03-05:48:03.918953] Error: ChecksumError: sha256 checksum failed for /tmp/root/spack-stage/spack-stage-aml-0.2.0-otgrqxle4zppijde5idaaeqq4jcfsrfn/aml-0.2.0.tar.gz
    Expected eea44780c11cff3f585f7ad1bc4fcc0d9d46aa8c27de234f6ae3703c9bb57684 but got 2044a2f3f1d7a19827dd9c0726172b690189b4d3fe938656c4160c022468cc4a
```

File: var/spack/repos/builtin/packages/aml/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/aml/package.py]:
```suggestion
    version('0.2.0', sha256='2044a2f3f1d7a19827dd9c0726172b690189b4d3fe938656c4160c022468cc4a')
```
You'll find this in the log of the failed GitLab CI job.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

**Comment by eugeneswalker** [comment]
We would love to include this in `E4S v22.02` slated to be finalized in the beginning of Feburary 2022.

**Comment by bernhardkaindl** [comment]
If I'm not mistaken, in any case, you'll have to fix the gitlab CI failure due to the still wrong sha256 for the 0.2.0 release:

https://gitlab.spack.io/spack/spack/-/jobs/1277246#L254

I pointed that out in an earlier comment too.

**Comment by perarnau** [comment]
@eugeneswalker I would really like that version to be included in the next E4S too. Right now this pull request disables the smoke test that was used for v0.1, as it does not really help much detecting issues with our variants. But if you prefer I can put it back and we can work out a better test later.

**Comment by eugeneswalker** [comment]
> 
Sounds good!

**Review by bernhardkaindl — DISMISSED** [review]
The new commit should fix the build. The PR can be merged after passing CI.

**Comment by bernhardkaindl** [comment]
@perarnau @eugeneswalker @NicolasDenoyelle Do need yet another fix before this update can be included in spack and E4S:

The package failed to build in the CI container which does not have libnuma installed:
1.  download https://gitlab.spack.io/spack/spack/-/jobs/1544550/artifacts/download
2. `unzip artifacts.zip`
3. `tail jobs_scratch_dir/logs/spack-build-out.txt`
```
checking numa.h usability... no
checking numa.h presence... no
checking for numa.h... no
configure: error: AML requires libnuma headers.
```
 You need to add a missing `depends_on` for the missing `libnuma`!
 
 Also test the resulting build recipe by installing the libnuma on your machine before running the build or build it in a fresh Ubuntu 18.04 installation, container or VM to verify that the build works on a clean system to prevent requiring yet another time and resouce-consuming PR push, review and CI/CD cycle, potentially causing you to miss your E4S integration goal.

File: var/spack/repos/builtin/packages/aml/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/aml/package.py:81]: @NicolasDenoyelle 
This fixes the build, but you have to apply this suggestion yourself since you denied access by maintainers like me to the branch of your PR:
```suggestion

    # - libnuma from the numactl package is required to build aml in GitLab CI
    depends_on('numactl')
```

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/aml/package.py:81]: The missing `depends_on` was added with commit 7bce013 below. This conversation is resolved therefore.

**Comment by bernhardkaindl** [comment]
Auto-merge is enabled now, it will be merged automatically after all tests and CI has passed, and the update is then included in spack and may be in E4S as far as I can see.

File: var/spack/repos/builtin/packages/aml/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/aml/package.py]: Why are these being commented out?

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Is there a reason for removing the current stand-alone tests?

If they don't work they should be updated; otherwise, they should be restored.

File: var/spack/repos/builtin/packages/aml/package.py
Comment (perarnau) [reply var/spack/repos/builtin/packages/aml/package.py]: Replaced with better test now.

File: var/spack/repos/builtin/packages/aml/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/aml/package.py]: Please move this below the `depends_on` directives.  

Having all of the directives -- `version`, `variant`, `depends_on`, etc. -- appear in the package above any methods is the convention used in Spack packages.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/aml/package.py]: By including the comment I meant that it needs to be split or updated since it lumps variants and configure options together.

Comment (perarnau) [reply var/spack/repos/builtin/packages/aml/package.py]: I see what you mean, should be better now

**Comment by perarnau** [comment]
@tldahlgren I believe I've pushed every change requested. Can you check again ?

**Review by perarnau — APPROVED** [review]

**Comment by NicolasDenoyelle** [comment]
I am handing this PR to someone else since I am no longer in charge of this package.
A similar PR will be opened later with a similar intent.

**Pull request closed without merging**
